---
title: "Access UKB data on Rosalind/CREATE"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to access UKB phenotype data on KCL Rosalind/CREATE HPC cluster}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    eval = FALSE,
    echo = TRUE,
    message = FALSE
)

requireNamespace("devtools", quietly = TRUE)
requireNamespace("stringr", quietly = TRUE)

devtools::load_all(".")
```

***

**Important. This document is only useful for UKB-approved KCL
reasearchers and their collaborators, with an account on the Rosalind/CREATE
HPC cluster.**

The ukbkings package works with a Rosalind/CREATE UKB project directory that
has been setup to use the package. See
[ukbproject](https://github.com/kenhanscombe/ukbproject) for setup
of the directory structure of a UKB project on Rosalind/CREATE.

***

<span style="color:dodgerblue;">**Contents:**</span>  
[0. tldr]  
[1. Preamble]  
[2. Installation]  
[3. Field subset file]  
[4. Read and write data]  
[5. Categorical codes]  
[6. Record level data]  
[7. Genetic data]

***

### 0. tldr

Install the package
```{R}
devtools::install_github("kenhanscombe/ukbkings", dependencies = TRUE, force = TRUE)
```

Write a serialised R dataframe to file for your required fields.
```{R}
bio_phen(
    project_dir = "<path_to_project_directory>",
    field = "<path_to_required_fields_file>", # one per line, no header
    out = "<stem_of_output_file>" # e.g. "data/ukb" writes "data/ukb.rds"
)
```

***Note**. `bio_phen` reads a withdrawal file from the project
directory and replaces phenotype values for to-be-excluded samples
with `NA`.*

<br>

Read the generated dataset into R.
```{R}
df <- readRDS("data/ukb.rds")
```

<br>

### 1. Preamble

Change to your user directory and make a project folder. Start an
interactive cluster session with sufficient memory to read the UKB data
(I think you can get away with less, but I've not worked that out yet).
Load the default cluster version of R 

From your study folder in your user director on Rosalind:
```{bash}
srun -p shared,brc --mem=30G --pty /bin/bash

module avail 2&>1 | grep R
module load <default_R_from_above>

R
```

<br>

Or, from your study folder in your user director on CREATE

```{bash}
srun -p cpu --time=0-1 --mem=30G --pty /bin/bash

module -r spider '^r$'
module load <default_R_from_above>

R
```

<br>

### 2. Installation

***Note**. All code blocks below are R, unless otherwise specified.*

<br>

Install from Github
```{R}
devtools::install_github("kenhanscombe/ukbkings", dependencies = TRUE, force = TRUE)
```

Load libraries. (I like to use tidyverse but it is not necessary, base
R is fine)
```{R}
library(ukbkings)
library(tidyverse)
```

Check help (press 'q' to exit).
```{R}
?ukbkings
```

***Note**. All included data and functionality is described in the
package index and also described on the
[ukbkings webpage](https://kenhanscombe.github.io/ukbkings/) under the
**Reference** tab.*

<br>

Check help on specific function, e.g.,
```{R}
?bio_phen
```

Point to the project directory.
```{R}
project_dir <- "<absolute_path_to_project_directory>"
```

<br>

### 3. Field subset file

You need a file with required fields, one per line, no header.

<br>

Read the project field-to-name "field finder" file, inspect the
variable metadata, and display the number of baskets included.
```{R}
f <- bio_field(project_dir)

head(f)
glimpse(f)

f %>%
    distinct(basket)
```

Search for variables required and add their field codes to a file, one
per line, no header. You can page through the file.
```{R}
f %>%
    select(name) %>%
    page(method = "print")
```

Or, search `name` column
```{R}
f %>%
    select(field, name) %>%
    filter(str_detect(name, "vegetables"))

f %>%
    select(field, name) %>%
    filter(str_detect(name, "ldl|triglycerides"))
```

Alternatively, search the
[UKB showcase](http://biobank.ndph.ox.ac.uk/showcase/) for a variable
of interest then filter on the `field` column in the field-to-name
dataframe (useful if multiple instances required). For example, if you
search for "cholesterol medication", the field stem you want is 6177.
```{R}
f %>%
    select(field, name) %>%
    filter(str_detect(field, "6177"))
```

`bio_field_add` is a convenience function for creating the one per line
required variables/fields file. By default the function appends fields.
Create the field subset file in your UKB user directory

```{R}
f %>%
    select(field, name) %>%
    filter(str_detect(field, "6177")) %>%
    bio_field_add("small_field_subset.txt")
```

Inspect the field selection file.
```{R}
system("cat small_field_subset.txt")
```

<br>

### 4. Read and write data

Read required fields and save as an rds file in your user directory.
Argument `out` should be a path to your UKB user directory
```{r}
bio_phen(
    project_dir,
    field = "small_field_subset.txt",
    out = "small_phenotype_subset"
)
```

***Note**. Dates in the UKB data are recorded in a variety of formats,
some of which are non-standard: "character string is not in a standard
unambiguous format", e.g., 2009-01-12T11:28:56. All date variables have
been left in character format for the user to convert as needed.*

<br>

Check the size of your file and read in your dataset
```{r}
system("ls -lh small_phenotype_subset.rds")
df <- readRDS("small_phenotype_subset.rds")
```

If required, rename columns from the default UKB field names to the
descriptive names used in the field-to-name "field finder" `name`
column.
```{R}
df <- bio_rename(df, f)
```

<br>

### 5. Categorical codes

Categorical field codings are included in the field finder.
```{R}
f %>%
    filter(field %in% names(df)) %>%
    select(field, categorical_coding)
```

Retrieve numerical "Value" and and associated "Meaning" for each
categorical code.
```{R}
cx <- bio_code(project_dir)
head(cx)
```

Look up a particular coding.
```{R}
cx %>%
    filter(Coding == 502)
```

<br>

### 6. Record level data

To query and read record-level data, use `bio_record` and
`bio_record_map`. List all record level data available for your project
```{r}
bio_record(project_dir)
```

<br>

You can retrieve the data as a [disk.frame](https://diskframe.com/index.html)
which you can inspect with functions like `head`, `names`, etc. This
data is still on disk and so does not require a large amount of
memory to read into R.

You can also pipe the disk.frame through dplyr verbs, e.g., for column
selection and row filtering. Use `collect` to read the results from disk
into an R dataframe. For example to retieve death data
```{r}
gp_scripts_diskf <- bio_record(project_dir, record = "gp_scripts")

gp_scripts_df <- gp_scripts_diskf %>%
    filter(str_detect(drug_name, "fluoxetine")) %>%
    select(eid, data_provider, issue_date) %>%
    collect()
```

Use the `subset` argument as a convenient way to read only the data for
the samples you're interested in, e.g. those with data on your phenotype
of interest. This will automatically read the data into a dataframe in
R - you do not need to `collect`.
```{R}
sample_subset <- c(321, 654, 987)

gp_scripts_df <- bio_record(
    project_dir,
    record = "gp_scripts",
    subset = sample_subset
)
```

<br>

To inspect several records at once use `bio_record_map`, which maps a
function to the data on disk without reading it into R. For example,
to quickly find which variables are in which tables
```{R}
bio_record_map(project_dir, func = names)
```

By default the function is mapped to all record table. You can also
specify the records you're interested in 
```{R}
bio_record_map(
    project_dir,
    func = head,
    records = c("gp_clinical", "gp_scripts", "gp_registrations")
)
```

<br>

For GP record details see UKB documentation:

* [Category 3001](http://biobank.ndph.ox.ac.uk/showcase/label.cgi?id=3001)
* [Resource 591: Primary Care Linked Data](http://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=591)
* [Resource 592: Clinical coding classification systems and maps](http://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=592)

<br>

For COVID-19 record details see UKB documentation:

* [Data field 40100](http://biobank.ndph.ox.ac.uk/showcase/field.cgi?id=40100)
* [COVID-19 test results data](http://biobank.ndph.ox.ac.uk/showcase/exinfo.cgi?src=COVID19_tests)
* [Resource 3151: Primary Care (GP) Data for COVID-19 Research](https://biobank.ndph.ox.ac.uk/ukb/refer.cgi?id=3151)

<br>

For HES in-patient record details see UKB documentation:

* [Category 2000](https://biobank.ndph.ox.ac.uk/showcase/label.cgi?id=2000)
* [Category 2006](https://biobank.ndph.ox.ac.uk/showcase/label.cgi?id=2006)
* [Resource 138483: Hospital Episode Statistics](https://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=138483)
* [Resource 141140: HES Data Dictionary](https://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=141140)

<br>

For death record details see UKB documentation:

* [Category 100093](https://biobank.ctsu.ox.ac.uk/crystal/label.cgi?id=100093)
* [Resource 134993: Health Outcomes Report](https://biobank.ctsu.ox.ac.uk/crystal/refer.cgi?id=134993)

<br>


### 7. Genetic data

For paths to the genetic data available for the project
```{r}
bio_gen_ls(project_dir)
```

For the genotyped data "sample information" files fam file and sample QC files

```{r}
bio_gen_fam(project_dir)
bio_gen_sqc(project_dir)
```


Read the relatedness data into a dataframe with `bio_gen_related`. To
get a dataframe of related samples to remove use
`bio_gen_related_remove` - uses [GreedyRelated](https://gitlab.com/choishingwan/GreedyRelated)
with default relatedness theshold, `thresh = 0.044`
```{r}
bio_gen_related(project_dir)
bio_gen_related_remove(project_dir)
```

To assign 1000 Genomes super population ancestry to your project-specific
pseudo-IDs.

```{r}
bio_gen_ancestry(project_dir)
```

To write a PLINK file of samples to `keep`, use
`bio_gen_write_plink_input`: `data` is either a vector of samples to
keep, or a dataframe with samples IDs to keep in the first column; `out`
is the file path to write output file to.

```{r}
bio_gen_write_plink_input(data, out)
```
